File: grocery/pipeline.py
"""The end-to-end extraction pipeline for one recipe input.

Kept free of any UI so it stays testable and reusable (the Streamlit app and,
later, the aggregation step both call it). It simply chains the stages:

    fetch/read/OCR (ingest) -> ingredient lines (extract) -> parse (CRF) -> categorize

All three input types (URL, document, photo) end up as a `RecipeContent` fed
through the same `Extractor` (`extract/base.py`) — a photo's OCR'd text and a
non-JSON-LD page's readable text both go through the same heuristic
line-finder (`extract/heuristic.py`).
"""

from grocery.extract.base import Extractor
from grocery.extract.categorize import categorize
from grocery.extract.ocr import extract_text as ocr_text
from grocery.extract.parse import parse_lines
from grocery.ingest import documents, web
from grocery.models import Ingredient, RecipeContent


def process_url(url: str, extractor: Extractor) -> tuple[RecipeContent, list[Ingredient]]:
    """Fetch and fully process one recipe URL into categorized ingredients.

    May raise `web.IngestError` if the page yields nothing usable.
    """
    content = web.fetch(url)
    lines = extractor.ingredient_lines(content)
    ingredients = parse_lines(lines, source_url=content.source_url)
    categorize(ingredients, extractor=extractor)
    return content, ingredients


def process_image(image: bytes, name: str, extractor: Extractor) -> tuple[RecipeContent, list[Ingredient]]:
    """Process a recipe *photo* into categorized ingredients.

    OCRs the image to text, then runs the same raw-text path a non-JSON-LD web
    page or document uses.
    """
    content = RecipeContent(title=name, raw_text=ocr_text(image))
    lines = extractor.ingredient_lines(content)
    ingredients = parse_lines(lines, source_url=None)
    categorize(ingredients, extractor=extractor)
    return content, ingredients


def process_document(data: bytes, name: str, extractor: Extractor) -> tuple[RecipeContent, list[Ingredient]]:
    """Process a recipe *document* (PDF/Word/text) into categorized ingredients."""
    content = RecipeContent(title=name, raw_text=documents.extract_text(data, name))
    lines = extractor.ingredient_lines(content)
    ingredients = parse_lines(lines, source_url=None)
    categorize(ingredients, extractor=extractor)
    return content, ingredients

File: android/core/src/main/kotlin/com/grocerylistbuilder/core/pipeline/Pipeline.kt
package com.grocerylistbuilder.core.pipeline

import com.grocerylistbuilder.core.config.Config
import com.grocerylistbuilder.core.extract.Categorizer
import com.grocerylistbuilder.core.extract.Extractor
import com.grocerylistbuilder.core.ingest.documents.DocumentTextExtractor
import com.grocerylistbuilder.core.ingest.ocr.OcrTextExtractor
import com.grocerylistbuilder.core.ingest.web.WebFetcher
import com.grocerylistbuilder.core.models.Ingredient
import com.grocerylistbuilder.core.models.RecipeContent

/**
 * The end-to-end extraction pipeline (mirrors grocery/pipeline.py). Free of any UI so it stays
 * testable and reusable, chaining the stages:
 *
 *     fetch/read (ingest) -> ingredients (extract) -> categorize
 */
object Pipeline {

    /** Fetch and process one recipe URL into categorized ingredients. May throw
     * [com.grocerylistbuilder.core.ingest.web.IngestError]. */
    suspend fun processUrl(url: String, webFetcher: WebFetcher, extractor: Extractor): Pair<RecipeContent, List<Ingredient>> {
        val content = webFetcher.fetch(url)
        val categorized = Categorizer.categorize(extractIngredients(content, extractor), Config.CATEGORIES, extractor)
        return content to categorized
    }

    /** Process an uploaded recipe document (PDF/Word/text) the same way as a non-JSON-LD URL. */
    suspend fun processDocument(
        data: ByteArray,
        name: String,
        documentTextExtractor: DocumentTextExtractor,
        extractor: Extractor,
    ): Pair<RecipeContent, List<Ingredient>> {
        val content = RecipeContent(title = name, rawText = documentTextExtractor.extractText(data, name))
        val categorized = Categorizer.categorize(extractIngredients(content, extractor), Config.CATEGORIES, extractor)
        return content to categorized
    }

    /** Process a recipe photo. A photo never has JSON-LD, so it always goes straight through
     * the CRF's combined extract+structure pass — there's no fast path to skip here. */
    suspend fun processImage(
        imageBytes: ByteArray,
        name: String,
        ocrTextExtractor: OcrTextExtractor,
        extractor: Extractor,
    ): Pair<RecipeContent, List<Ingredient>> {
        val text = ocrTextExtractor.recognizeText(imageBytes)
        val content = RecipeContent(title = name, rawText = text)
        val categorized = Categorizer.categorize(extractor.structureFromText(text), Config.CATEGORIES, extractor)
        return content to categorized
    }

    /**
     * JSON-LD lines: wrapped verbatim, zero added latency (unchanged since Phase 1). Raw text
     * (OCR/document/non-JSON-LD page): the CRF's combined extract+structure pass. Deliberate
     * default — JSON-LD is the common case (most recipe sites have it) and is already clean and
     * free; running it through the line-finder too would risk mangling already-clean names for no
     * benefit. [Combiner]'s exact-name matching still merges JSON-LD ingredients fine without
     * splitting them.
     */
    private suspend fun extractIngredients(content: RecipeContent, extractor: Extractor): List<Ingredient> {
        val lines = extractor.ingredientLines(content)
        if (lines.isNotEmpty()) return lines.map { line -> Ingredient(name = line, sourceUrl = content.sourceUrl) }
        val text = content.rawText.trim()
        return if (text.isEmpty()) emptyList() else extractor.structureFromText(text)
    }
}
